### ISSUE TITLE
Undo Partitioning: Simple Time Based

### ISSUE THREAD

**Issue by julianopg** [keithf4/pg_partman#193]
Hi Keithf4,

I did a default partitioning (part_config:  premake = '4', optimize_trigger='4') in mydatabase and now I would like to have more months, so, I have increased these parameter: _UPDATE partman.part_config SET premake = '10', optimize_trigger='10' WHERE parent_table = 'myschema.mytable'_

The problem was just months before was created and months after wasn't!

So, I have undo the partitioning to redo with a new value using the command: _python undo_partition.py -p myschema.mytable -c host=localhost -t time_, but I have got the following error:
**Traceback (most recent call last):
  File "undo_partition.py", line 262, in <module>
    partman_schema = get_partman_schema(conn)
  File "undo_partition.py", line 40, in get_partman_schema
    partman_schema = "\"" + cur.fetchone()[0] + "\""
TypeError: 'NoneType' object has no attribute '__getitem__'**

I have fixed it add the parameter _"-c dbname=mydatabase"_ to connect to my database: _python undo_partition.py -p myschema.mytable -c host=localhost -c dbname=mydatabase -t time_, but I have got the following error:
**Attempting to turn off autovacuum for partition set...
	... Success!
Attempting to reset autovacuum for old parent table...
	... Success!
Traceback (most recent call last):
  File "undo_partition.py", line 269, in <module>
    total = undo_partition_data(conn, partman_schema)
  File "undo_partition.py", line 194, in undo_partition_data
    cur.execute(query, li)
psycopg2.InternalError: No non-native configuration found for pg_partman for given parent table: myschema.mytable
CONTEXT: 
DETAIL: 
HINT:** 

I don't like to lose the data in these tables, can you help me please?

**Comment by julianopg** [comment]
Just to let you know I'm using Postgres 9.4 and Ubuntu 16:
PostgreSQL 9.4.12 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609, 64-bit

And I have installed partman from the Postgres official repository by the command: 
sudo apt-get install postgresql-9.4-partman python-psycopg2

**Comment by keithf4** [comment]
Can you please provide the schema of the parent table (\d+ in psql) as well as the entire contents of the part_config table for the given partition set 
```
SELECT * FROM partman.part_config WHERE parent_table = 'myschema.mytable';
```

**Comment by keithf4** [comment]
Can you also check which version of pg_partman is actually installed? I do not maintain the package in the repository. Just do \dx+ in psql.

**Comment by julianopg** [comment]
Hi

Commands and results bellow:

**\d+myschema.mytable**

                                        Table "myschema.mytable"
           Column           |           Type           |  Modifiers   | Storage  | Stats target | Description 
----------------------------+--------------------------+--------------+----------+--------------+-------------
 id                     	| text                     |              | extended |              | 
 type_id                	| integer                  |              | plain    |              | 
 time_stamp             	| timestamp with time zone | not null     | plain    |              | 
 column4		       | boolean                  | default true | plain    |              | 
Indexes:
    "mytable_id_type_id_idx" btree (id, type_id)
    "mytable_type_id_idx" btree (type_id)
    "mytable_time_stamp_idx" btree (time_stamp)
Triggers:
    mytable_part_trig BEFORE INSERT ON myschema.mytable FOR EACH ROW EXECUTE PROCEDURE myschema.mytable_part_trig_func()
Child tables: myschema.mytable_p2017_03,
              myschema.mytable_p2017_04,
              myschema.mytable_p2017_05,
              myschema.mytable_p2017_06,
              myschema.mytable_p2017_07,
              myschema.mytable_p2017_08,
              myschema.mytable_p2017_09,
              myschema.mytable_p2017_10,
              myschema.mytable_p2017_11


**\dx**
pg_partman   | 3.0.2   | partman    | Extension to manage partitioned tables by time or ID


**SELECT * FROM partman.part_config WHERE parent_table = 'myschema.mytable';**

 parent_table | control | partition_type | partition_interval | constraint_cols | premake | optimize_trigger | optimize_constraint | epoch | inherit_fk | retention | retention_schema | retention_keep_table | retention_keep_index | infinite_time_partitions | datetime_string | automatic_maintenance | jobmon | sub_partition_set_full | undo_in_progress | trigger_exception_handling | upsert | trigger_return_null 
--------------+---------+----------------+--------------------+-----------------+---------+------------------+---------------------+-------+------------+-----------+------------------+----------------------+----------------------+--------------------------+-----------------+-----------------------+--------+------------------------+------------------+----------------------------+--------+---------------------
(0 rows)

**select partman.check_parent();**

 check_parent 
--------------
(0 rows)


Looks like the parents is gone, is it possible overwrite?

**Comment by keithf4** [comment]
I'd done some testing before you provided the above info. Nothing has changed in my findings based on that and so far I can only repeat what you're seeing with additional future partitions not being created if no current data (as of the partition column) exists in the table. As soon as current data exists in the partition set, then the new premake value goes into affect the next maintenance run.
```
keith=# \dx
                               List of installed extensions
    Name    | Version |   Schema   |                     Description                      
------------+---------+------------+------------------------------------------------------
 pg_partman | 3.0.2   | partman    | Extension to manage partitioned tables by time or ID
 pgtap      | 0.98.0  | public     | Unit testing for PostgreSQL
 plpgsql    | 1.0     | pg_catalog | PL/pgSQL procedural language
(3 rows)

keith=# create table mytable (id bigint not null, updated_at timestamptz not null);
CREATE TABLE
keith=# \d+ mytable 
                                  Table "public.mytable"
   Column   |           Type           | Modifiers | Storage | Stats target | Description 
------------+--------------------------+-----------+---------+--------------+-------------
 id         | bigint                   | not null  | plain   |              | 
 updated_at | timestamp with time zone | not null  | plain   |              | 

keith=# select partman.create_parent('public.mytable', 'updated_at', 'partman', 'monthly');
 create_parent 
---------------
 t
(1 row)

keith=# \d+ mytable
                                  Table "public.mytable"
   Column   |           Type           | Modifiers | Storage | Stats target | Description 
------------+--------------------------+-----------+---------+--------------+-------------
 id         | bigint                   | not null  | plain   |              | 
 updated_at | timestamp with time zone | not null  | plain   |              | 
Triggers:
    mytable_part_trig BEFORE INSERT ON mytable FOR EACH ROW EXECUTE PROCEDURE mytable_part_trig_func()
Child tables: mytable_p2017_03,
              mytable_p2017_04,
              mytable_p2017_05,
              mytable_p2017_06,
              mytable_p2017_07,
              mytable_p2017_08,
              mytable_p2017_09,
              mytable_p2017_10,
              mytable_p2017_11

keith=# UPDATE partman.part_config SET premake = '10', optimize_trigger='10' WHERE parent_table = 'public.mytable';
UPDATE 1
keith=# select partman.run_maintenance();
 run_maintenance 
-----------------
 
(1 row)
```
You can see here that new future partitions were not created. This is because the partition set has no current data. As soon as I insert a current date value, maintenance catches up with the new premake value.
```
keith=# \d+ mytable
                                  Table "public.mytable"
   Column   |           Type           | Modifiers | Storage | Stats target | Description 
------------+--------------------------+-----------+---------+--------------+-------------
 id         | bigint                   | not null  | plain   |              | 
 updated_at | timestamp with time zone | not null  | plain   |              | 
Triggers:
    mytable_part_trig BEFORE INSERT ON mytable FOR EACH ROW EXECUTE PROCEDURE mytable_part_trig_func()
Child tables: mytable_p2017_03,
              mytable_p2017_04,
              mytable_p2017_05,
              mytable_p2017_06,
              mytable_p2017_07,
              mytable_p2017_08,
              mytable_p2017_09,
              mytable_p2017_10,
              mytable_p2017_11

keith=# insert into mytable values (1, now());
INSERT 0 0
keith=# select partman.run_maintenance();
 run_maintenance 
-----------------
 
(1 row)


keith=# \d+ mytable
                                  Table "public.mytable"
   Column   |           Type           | Modifiers | Storage | Stats target | Description 
------------+--------------------------+-----------+---------+--------------+-------------
 id         | bigint                   | not null  | plain   |              | 
 updated_at | timestamp with time zone | not null  | plain   |              | 
Triggers:
    mytable_part_trig BEFORE INSERT ON mytable FOR EACH ROW EXECUTE PROCEDURE mytable_part_trig_func()
Child tables: mytable_p2017_03,
              mytable_p2017_04,
              mytable_p2017_05,
              mytable_p2017_06,
              mytable_p2017_07,
              mytable_p2017_08,
              mytable_p2017_09,
              mytable_p2017_10,
              mytable_p2017_11,
              mytable_p2017_12,
              mytable_p2018_01,
              mytable_p2018_02,
              mytable_p2018_03,
              mytable_p2018_04,
              mytable_p2018_05
```
Also, undoing seems to work without issue as well. If you're giving more than one parameter to the `-c` option, be sure they're quoted as seen below.
```
$ python bin/undo_partition.py -p public.mytable -c "host=localhost dbname=keith" -t time
Attempting to turn off autovacuum for partition set...
	... Success!
Partitions undone this batch: 5. Rows undone this batch: 1
Total rows moved: 1
Partitions undone this batch: 10. Rows undone this batch: 0
Total rows moved: 1
Total partitions undone: 15
Running vacuum analyze on parent table...
Attempting to reset autovacuum for old parent table...
	... Success!

keith=# \d+ mytable
                                  Table "public.mytable"
   Column   |           Type           | Modifiers | Storage | Stats target | Description 
------------+--------------------------+-----------+---------+--------------+-------------
 id         | bigint                   | not null  | plain   |              | 
 updated_at | timestamp with time zone | not null  | plain   |              | 

keith=# select * from mytable;
 id |          updated_at           
----+-------------------------------
  1 | 2017-07-24 08:26:05.076856-04
(1 row)
```
Also be aware that you did not specify the --droptable option to the undo_partition.py script, so some old child tables may still exist, but are not part of the partition set. That means if you try and redo this partition set again, you may run into the following error:
```
keith=# select partman.create_parent('public.mytable', 'updated_at', 'partman', 'monthly');
WARNING:  No child tables created. Check that all child tables did not already exist and may not have been part of partition set. Given parent has still been configured with pg_partman, but may not have expected children. Please review schema and config to confirm things are ok.
 create_parent 
---------------
 f
(1 row)
try
I did the same thing and you can see it threw that warning because all my child tables still exist. The default of the script is to be the least destructive as possible.
keith=# \dt
                   List of relations
 Schema |             Name              | Type  | Owner 
--------+-------------------------------+-------+-------
 public | mytable                       | table | keith
 public | mytable_p2017_03              | table | keith
 public | mytable_p2017_04              | table | keith
 public | mytable_p2017_05              | table | keith
 public | mytable_p2017_06              | table | keith
 public | mytable_p2017_07              | table | keith
 public | mytable_p2017_08              | table | keith
 public | mytable_p2017_09              | table | keith
 public | mytable_p2017_10              | table | keith
 public | mytable_p2017_11              | table | keith
 public | mytable_p2017_12              | table | keith
 public | mytable_p2018_01              | table | keith
 public | mytable_p2018_02              | table | keith
 public | mytable_p2018_03              | table | keith
 public | mytable_p2018_04              | table | keith
 public | mytable_p2018_05              | table | keith
```
If run into this and want to keep these tables as part of the partition set, you can just use the ALTER TABLE ... INHERIT... command.
```
ALTER TABLE mytable_p2017_07 INHERIT mytable;
```
How the child tables are still part of the inheritance set, but the entry is missing from part_config, I'm really not sure. If everything looks fine in the parent and child tables themselves, you can just insert the entry into part_config manually. Easiest thing to do would be to recreate another partition set with a similar setup to get all the values and then create an INSERT statement based on that for your old table. 

Nothing I've seen you do should have destroyed any data, but I'm not sure where exactly it may reside at this time. It'll either be in the parent, the children (some of which may have been uninherited) or a mixture of both.

**Comment by julianopg** [comment]
Hi,

Thanks for the explanation about the future data in the table, that explains a lot

I have dropped the extension, stopped the postgres, restarted the machine and recreated the extension.

So, I have executed the maintenance manually:
SELECT partman.run_maintenance();

And finally the parent is back:
SELECT partman.check_parent();
 check_parent            
-----------------------------------
 (myschema.mytable,58781531)
(1 row)

SELECT * from partman.part_config;
      parent_table      |  control   | partition_type | partition_interval | constraint_cols | premake | optimize_trigger | optimize_constraint | epoch | inherit_fk | retention | retention_schema | retention_keep_table | retention_keep_index | infinite_time_partitions | datetime_string | automatic_maintenance | jobmon | sub_partition_set_full | undo_in_progress | trigger_exception_handling | upsert | trigger_return_null 
------------------------+------------+----------------+--------------------+-----------------+---------+------------------+---------------------+-------+------------+-----------+------------------+----------------------+----------------------+--------------------------+-----------------+-----------------------+--------+------------------------+------------------+----------------------------+--------+---------------------
 myschema.mytable | time_stamp | partman        | 1 mon              |                 |       4 |                4 |                  30 | none  | t          |           |                  | t                    | t                    | f                        | YYYY_MM         | on                    | t      | f                      | f                | f                          |        | t
(1 row)

Just for check it I have tried to alter the table and I got as you said:
ALTER TABLE mytable_p2017_07 INHERIT mytable;
relation "mytable" would be inherited from more than once

So the python script is fine now
python undo_partition.py -p myschema.mytable -c "host=localhost dbname=cc_reporting" -t time
Attempting to turn off autovacuum for partition set...
	... Success!
Partitions undone this batch: 9. Rows undone this batch: 0
Total rows moved: 0
Total partitions undone: 9
Running vacuum analyze on parent table...
Attempting to reset autovacuum for old parent table...
	... Success!

 \d+ myschema.mytable
The child table is gone and the data looks good in the main table
I will recreate the partitioning once more with the new values

 Thanks so much for your help I really appreciate it!
